### DISCUSSION TITLE
How to recuntruct image back with `invtransform` to get `np.close()` to the original image?

### DISCUSSION THREAD

**Discussion by kondratevakate** [ANTsX/ANTs#1321] [Q&A]
Hi all! I want to resample images from the heterogeneous dataset to equal voxel spacing [1,1,1]. Yet I want to save the transformation to do `invtransform` and get **the same image** by  `np.allclose()` . How can I do that?

### 1. I found that if I will resample the image to the original spacing, I will not get the original image (it will be considerable MAE):
```

img = ants.image_read(test_path + 'T1.nii.gz')
img_res = ants.resample_image(img, (1, 1, 1), False, 4)
img_res_reverse = ants.resample_image(img_res, img.spacing, False, 4)
np.allclose(img_res_reverse.numpy(), img.numpy())
```
**False**
```
mae = np.sum(np.absolute(img_res_reverse.numpy() - img.numpy()))
mae, img_res_reverse.numpy().mean(),  img.numpy().mean()
```
**(1987788.2, 172.29662, 172.29372)**

### 2. I found, that if I will do it with `registration`, the error will be still huge:

```
img_res = ants.resample_image(img, (1, 1, 1), False, 0)
resample = ants.registration(fixed=img_res, moving=img, type_of_transform='Rigid')
img_res = ants.apply_transforms(img_res, img, transformlist = resample['fwdtransforms'][0])
img_res_reverse = ants.apply_transforms(img, img_res, transformlist = resample['invtransforms'][0])
mae = np.sum(np.absolute(img_res_reverse.numpy() - img.numpy()))
mae, img_res_reverse.numpy().mean(),  img.numpy().mean()
```
**(9012015.0, 172.40034, 172.29372)**

How one can do the resampling transformation in ‘ants’ in order to get ~ the same image by applying inverse transform?

**Comment by hjmjohnson** [comment]
The process you describe is an inherently lossy process.

As a thought experiment, imagine what happens if you were to resample to voxel space with 1000 x 1000 x 1000mm (instead of a 1 x 1 x 1 mm) image.   The entire original image would fit in 1 voxel of the resampled space and have a single value.  Then when you go back to the original space, there is only 1 value to interpolate from.

**Reply by kondratevakate** [reply]
Thank you, in the case of image interpolation it might be not that relevant. Yet in the case of segmentation and data lebelling, it is important to minimize approximation error. I will probably explain this effect with the code below

**Comment by kondratevakate** [comment]
![image](https://user-images.githubusercontent.com/34157226/158602929-3bb4dd85-d17f-40e2-b0ef-1776d6f9013c.png)

On the first row, I have image segmentation labels for BraTS2021, on the left - original image. And on the right - image after resampling and then resampling back.  
Then I take only class 1 on the next row. On the left - original class 1, on the left - resampled and resampled back, and then rounded.
And I have this "corona effect" which spoils the label after reconstruction, which can be explained by interpolation. 

That is kinda strange and maybe caused by contamination with other labels. Yet, still why even for labels I was not able to get accurate inverse reconstruction?  

Thanks!

```img = ants.image_read(path + '/' + image +'/' + 'CT1_SEG.nii.gz')
# resample
img_res = ants.resample_image(img, (1, 1, 1), False, 0)
# apply inverse transform
img_res_reverse = ants.resample_image(img_res, img.spacing, False, 0)
# round inverse labels
pred = np.round(img_res_reverse.numpy(), 0)
# calculate metrics
distances = metrics.compute_surface_distances((img[:,:,:]==1), (pred[:,:,:]==1), spacing)
dice = metrics.compute_dice_coefficient((img[:,:,:]==1), (pred[:,:,:]==1))
s_dice = metrics.compute_surface_dice_at_tolerance(distances, 1)```

** Dice  0.7125198227137803 S Dice 0.3386129078508565**
